lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs

```diff
@@ -0,0 +1,173 @@
+    let
+        blockHash =
+            Hash {getHash = "\216OY\rX\199\234\188.<O\\\244Y\211\210\254\224`i\216\DC3\167\132\139\154\216\161T\174\247\155"}
+        tl = newTransactionLayer blockHash :: TransactionLayer (Jormungandr n)
+        calcSize = estimateSize tl sel
+    in calcSize === Quantity 0
+
+instance Arbitrary CoinSelection where
+    shrink sel@(CoinSelection inps outs chgs) = case (inps, outs, chgs) of
+        ([_], [_], []) ->
+            []
+        _ ->
+            let
+                inps' = take (max 1 (length inps `div` 2)) inps
+                outs' = take (max 1 (length outs `div` 2)) outs
+                chgs' = take (length chgs `div` 2) chgs
+                inps'' = if length inps > 1 then drop 1 inps else inps
+                outs'' = if length outs > 1 then drop 1 outs else outs
+                chgs'' = drop 1 chgs
+            in
+                filter (\s -> s /= sel && isValidSelection s)
+                    [ CoinSelection inps' outs' chgs'
+                    , CoinSelection inps' outs chgs
+                    , CoinSelection inps outs chgs'
+                    , CoinSelection inps outs' chgs
+                    , CoinSelection inps'' outs'' chgs''
+                    , CoinSelection inps'' outs chgs
+                    , CoinSelection inps outs'' chgs
+                    , CoinSelection inps outs chgs''
+                    ]
+    arbitrary = do
+        outs <- choose (1, 10)
+            >>= \n -> vectorOf n arbitrary
+            >>= genTxOut
+        genSelection (NE.fromList outs)
+
+deriving instance Arbitrary a => Arbitrary (ShowFmt a)
+
+-- Check whether a selection is valid
+isValidSelection :: CoinSelection -> Bool
+isValidSelection (CoinSelection i o c) =
+    let
+        oAmt = sum $ map (fromIntegral . getCoin . coin) o
+        cAmt = sum $ map (fromIntegral . getCoin) c
+        iAmt = sum $ map (fromIntegral . getCoin . coin . snd) i
+    in
+        (iAmt :: Integer) >= (oAmt + cAmt)
+
+genTxOut :: [Coin] -> Gen [TxOut]
+genTxOut coins = do
+    let n = length coins
+    outs <- vectorOf n arbitrary
+    return $ zipWith TxOut outs coins
+
+genSelection :: NonEmpty TxOut -> Gen CoinSelection
+genSelection outs = do
+    let opts = CS.CoinSelectionOptions 100
+    utxo <- vectorOf (NE.length outs * 3) arbitrary >>= genUTxO
+    case runIdentity $ runExceptT $ largestFirst opts outs utxo of
+        Left _ -> genSelection outs
+        Right (s,_) -> return s
+
+genUTxO :: [Coin] -> Gen UTxO
+genUTxO coins = do
+    let n = length coins
+    inps <- vectorOf n arbitrary
+    outs <- genTxOut coins
+    return $ UTxO $ Map.fromList $ zip inps outs
+
+instance Arbitrary Coin where
+    shrink (Coin c) = Coin <$> shrink (fromIntegral c)
+    arbitrary = Coin <$> choose (1, 200000)
+
+instance Arbitrary Address where
+    shrink _ = []
+    arbitrary =
+        pure $ Address "\131\&3$\195xi\193\"h\154\&5\145}\245:O\"\148\163\165/h^\ENQ\245\248\229;\135\231\234E/"
```

Comment:
probably will need to be not so specific here, for the test I added seems not forbidden